crates/paths/src/utils.rs

```diff
@@ -0,0 +1,122 @@
+use std::path::{Path, PathBuf};
+
+pub(crate) trait PathBufExt {
+    fn joined<P: AsRef<Path>>(self, path: P) -> Self;
+    fn joined_int<I: itoa::Integer>(self, path_seg: I) -> Self;
+}
+
+impl PathBufExt for PathBuf {
+    fn joined<P: AsRef<Path>>(mut self, path: P) -> Self {
+        self.push(path);
+        self
+    }
+
+    fn joined_int<I: itoa::Integer>(self, path_seg: I) -> Self {
+        self.joined(itoa::Buffer::new().format(path_seg))
+    }
+}
+
+/// Declares a new strongly-typed path newtype.
+///
+/// ```
+/// # use spacetimedb_paths::path_type;
+/// path_type! {
+///     /// optional docs
+///     // optional. if false, makes the type's constructor public.
+///     #[non_exhaustive(FALSE)]
+///     FooPath: dir // or file. adds extra utility methods for manipulating the file/dir
+/// }
+/// ```
+#[macro_export]
+macro_rules! path_type {
+    ($(#[doc = $doc:literal])* $(#[non_exhaustive($non_exhaustive:tt)])? $name:ident) => {
+        $(#[doc = $doc])*
+        #[derive(Clone, Debug, $crate::__serde::Serialize, $crate::__serde::Deserialize)]
+        #[serde(transparent)]
+        #[cfg_attr(all($($non_exhaustive)?), non_exhaustive)]
+        pub struct $name(pub std::path::PathBuf);
```

Comment:
nit: I don't love the `.0` pattern for unwrapping these paths. Imo a `.untyped()` or `.to_pathbuf()` or something would be clearer.